tests/css_indent_styles/dump.css

```diff
@@ -0,0 +1,52 @@
+.container,
+table {
+  background: papayawhip;
+}
+
+.container-cell {
+  background: bisque;
+}
+
+.container-cell--highlight {
+  background: lime;
+}
+
+.container-bluff {
+  background: firebrick;
+}
+
+table {
+  background: papayawhip;
+}
+
+table tr {
+  background: palegreen;
+}
+
+table tr td {
+  background: palegoldenrod;
+}
+
+table tr.dancer {
+  background: palevioletred;
+}
+
+table train {
+  background: papayawhip;
+}
+
+div.dancer {
+  background: papayawhip;
+}
+
+div.dancer {
+  table {
+    tr {
+      background: papayawhip;
+
+      td {
+        background: papayawhip;
+      }
+    }
+  }
+}
```

Comment:
I would like to see crazy SCSS and Less tests as well.

- What happens when an indented rule has nested rules?
- What happens when SCSS/Less/PostCSS interpolations are used in selectors?
- What about pseudo-classes (`:hover`) and pseudo-elements (`::after`)?
- What about selectors with commas? `table th, table td`?

And some case insensitiviy stuff. I guess `TABLE TR` should be indented under `table`?

Reply:
Awesome, thanks @lydell! I'll try to address your feedback over the next few days. Commas are already supported, and the [Bootstrap output CSS formatting example has some examples](https://gist.github.com/holloway/66cb1585cc11ac5c96848dfcca5d1c5c#file-bootstrap-with-indenting-css-L150)  (and of course that's just some arbitrary CSS I chose, I don't imagine that people would normally format their output).

Just to add another question to the list:

* What happens with code comments?

Reply:
For performance testing is there a tool that Prettier devs usually use? If there's nothing currently I could use my [whats-the-damage](https://www.npmjs.com/package/whats-the-damage) tool if that's ok... it can measure durations, CPU, and memory usage.

Reply:
Don't focus to much on the performance thing. It's enough to just run the CLI on a big piece of CSS with and without the option to see if there's a huge difference. For example, a mistake in the selector-checking loop could cause a massive slowdown, but if it's all correct we should be fine.

Also not that my questions in the bullet points in my above comments were intended more to be requests for additions to the tests, rather than just questions. Sorry for being unclear.

Reply:
I've just pushed a new version with CLI mode, comment indenting, nested styles, and more tests. The indenting or passthrough code is now a single line of code too.

For nested styles the code skips indentation changes for that block, but it retains any indentation from the outer selector. It's hard to explain but I think it looks right,

before:
```css
div {
  background: papayawhip;
}

div s {
  background: palegreen;
}

div s.ghost {
  background: palegoldenrod;

  .coast {
      background-color: gray;
  }
}

div s.dancer {
  // this should be indented with previous nested style
  background: palevioletred;
}

div section {
  // this shouldn't be indented with "div s" because although it starts with the same characters
  // we treat element name as atomic.
  background: papayawhip;
}
```

after
```css
div {
  background: papayawhip;
}

  div s {
    background: palegreen;
  }

    div s.ghost {
      background: palegoldenrod;

      .coast {
        background-color: gray;
      }
    }

    div s.dancer {
      // this should be indented with previous nested style
      background: palevioletred;
    }

  div section {
    // this shouldn't be indented with "div s" because although it starts with the same characters
    // we treat element name as atomic.
    background: papayawhip;
  }
```

So for `div s.ghost` that outer selector for the nested style is used for indenting, but the inner ones are ignored and they just use the existing Prettier formatting. If it didn't skip the `.coast` then I'm not sure what would be expected...the code might have to resolve the full selector and perhaps resolve `&` and so on, which (I think) would make the code far too complex.



The comment indenting looks like this,

before
```css
a {
    text-decoration: none;
}
// a comment that should be indented at same column as following selector
a:hover {
    text-decoration: underline overline; 
}
```

after
```css
a {
  text-decoration: none;
}
  // a comment that should be indented at same column as following selector
  a:hover {
    text-decoration: underline overline;
  }
```

This way if there's an indentation shape any comments will adhere to it. This example also shows pseudo classes, and pseudo elements work the same way.

For performance just using `time` on the command line it's averaging 0.254s without indent-styles and 0.267s with.

> And some case insensitiviy stuff. I guess TABLE TR should be indented under table

This makes sense per-spec for HTML, but I suppose there are extremely niche cases like [XML CSS](http://www.xmlplease.com/xml/xmlcss/) that afaik are case-sensitive. More popular would be [Attribute values](https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors) which seem to be case-sensitive (`a[data-amiga="cOmMoDoRe"]` !== a[data-amiga="CoMmOdOrE"]`) unless there's an `i` flag. These per-spec rules are somewhat complex, and I wonder if people writing `TABLE TR` followed by `table tr td` is also a niche case to support. Maybe this aspect of the code should be left as-is to keep the code simpler in `getSelectorPattern`, even if it is wrong per-spec. I don't have a strong opinion here though so I'll make elements insensitive with the next PR revision unless you suggest otherwise.

TODO:

* More tests for every CSS node type appearing in the file.
* Confirm whether any of these nodes should be indented too. I'm hoping not because I wouldn't want to affect much more of the main code-paths. The current test file has a Sass variable declaration between two selectors 
* "What happens when SCSS/Less/PostCSS interpolations are used in selectors?" 
* Anything else?

I'll try to get a new version in the next week.

Thanks so much Lydell!

Reply:
Re. `.coast`: That's the expected output. I wonder if it is possible to trick the indenter into indenting too much if you use something like this?

```css
div {
  div.foo {}
}
div.foo.bar {}
```

The expected output is:

```css
div {
  div.foo {}
}
  div.foo.bar {}
```

and not:

```css
div {
  div.foo {}
}
    div.foo.bar {} // A mistake in the logic could cause double indentation here. 
```

Re. performance: Great, nothing to worry about there, I guess.

Re. case insensitivity: I totally understand that it is an edge case, but Prettier already lowercases stuff, so we _have_ to do it in order to make the formatting stable. We don't want this to happen:

```css
TABLE {}
table.zebra {}
```
→
```css
table {}
table.zebra {}
```
→
```css
table {}
  table.zebra {}
```

Reply:
Thanks, so I'll add that indenter trick as a test (currently the indenter passes that test).

For case insensitivity.. that makes a lot of sense, so I'll definitely do it and try to adhere to the same normalization rules that Prettier already has.